### ISSUE TITLE
Connection Closed by the other side error when using Kubernetes API 

### ISSUE THREAD

**Issue by Anumita** [hazelcast/hazelcast-kubernetes#299]
I'm having some issues while using kubernetes Api discovery mode. My setup is with an embedded cache server, a kubernetes Deployment with just two replicas. Error in the hazelcast sidecar: Getting INFO: [172.10.0.18]:5701 [dev] [3.12.11] Connection[id=1, /127.0.0.1:5701->/127.0.0.1:47194, qualifier=null, endpoint=null, alive=false, type=NONE] closed. Reason: Connection closed by the other side error. My service yaml is 

```
apiVersion: v1
kind: Service
metadata:
  name: go-hello-world-sidecar
  labels:
    app: go-hello-world-sidecar
spec:
  type: ClusterIP
  clusterIP: 10.96.0.3
  ports:
  - name: http
    port: 9090
    targetPort: 8080
  - name: hazelcast
    port: 5701
  - name: promhttp
    port: 8443
  selector:
    app: go-hello-world-sidecar
status:
  loadBalancer: {}
```

If i don't specify a clusterIP , the members are able to discover each other. This error is coming only when i specify a clusterIP.
The hazelcast config yaml:
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: hazelcast-configuration
data:
  hazelcast.yaml: |-
    hazelcast:
      network:
        join:
          multicast:
            enabled: false
          kubernetes:
            enabled: true
            namespace: kube-system
            service-name: go-hello-world-sidecar
            kubernetes-api-retries: 15
      partition-group:
        enabled: true
        group-type: NODE_AWARE
```
Not sure why this is happening only with ClusterIP type with predefined clusterIP. Am I missing something? 
The hazelcast image is 3.12.11 and kubernetes version is 1.18.15

**Comment by hasancelik** [comment]
Hi @Anumita,

I have tried to create similar env at GKE based on information that you provided but I could not reproduce your issue. Here is the my YAML files:

```
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: hazelcast-sidecar
  labels:
    app: hazelcast-sidecar
spec:
  replicas: 2
  serviceName: hazelcast-sidecar-service
  selector:
    matchLabels:
      app: hazelcast-sidecar
  template:
    metadata:
      labels:
        app: hazelcast-sidecar
    spec:
      containers:
        - name: hazelcast
          image: hazelcast/hazelcast:3.12.11
          ports:
            - name: hazelcast
              containerPort: 5701
          volumeMounts:
            - name: hazelcast-storage
              mountPath: /data/hazelcast
          env:
            - name: JAVA_OPTS
              value: "-Dhazelcast.rest.enabled=true -Dhazelcast.config=/data/hazelcast/hazelcast.yaml"

        - name: sidecar-container
          image: busybox
          command: ["/bin/sh"]
          args: ["-c", "while true; do echo echo $(date -u) 'Hi I am from Sidecar container' >> /var/log/index.html; sleep 5;done"]

      volumes:
        - name: hazelcast-storage
          configMap:
            name: hazelcast-configuration

---

apiVersion: v1
kind: Service
metadata:
  name: hazelcast-sidecar
spec:
  type: ClusterIP
  clusterIP: 10.168.0.3
  selector:
    app: hazelcast-sidecar
  ports:
    - name: hazelcast
      port: 5701
```

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: hazelcast-configuration
data:
  hazelcast.yaml: |-
    hazelcast:
      network:
        join:
          multicast:
            enabled: false
          kubernetes:
            enabled: true
            namespace: default
            service-name: hazelcast-sidecar
```
Cluster state:
```
$ kg all

NAME                      READY   STATUS    RESTARTS   AGE
pod/hazelcast-sidecar-0   2/2     Running   0          9m6s
pod/hazelcast-sidecar-1   2/2     Running   0          9m3s

NAME                        TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)    AGE
service/hazelcast-sidecar   ClusterIP   10.168.0.3   <none>        5701/TCP   9m7s
service/kubernetes          ClusterIP   10.168.0.1   <none>        443/TCP    3h49m

NAME                                 READY   AGE
statefulset.apps/hazelcast-sidecar   2/2     9m9s
```

Member log:
```
$ klo pod/hazelcast-sidecar-0 hazelcast
....
INFO: [10.164.1.10]:5701 [dev] [3.12.11] Kubernetes Discovery properties: { service-dns: null, service-dns-timeout: 5, service-name: hazelcast-sidecar, service-port: 0, service-label: null, service-label-value:true, namespace: default, pod-label: null, pod-label-value: null, resolve-not-ready-addresses: false, use-node-name-as-external-address: false, kubernetes-api-retries: 3, kubernetes-master: https://kubernetes.default.svc}
Feb 24, 2021 11:55:40 AM com.hazelcast.spi.discovery.integration.DiscoveryService
INFO: [10.164.1.10]:5701 [dev] [3.12.11] Kubernetes Discovery activated with mode: KUBERNETES_API
....
INFO: [10.164.1.10]:5701 [dev] [3.12.11]
Members {size:2, ver:2} [
        Member [10.164.2.6]:5701 - 19148d83-a732-4027-957d-ae9148aa4046
        Member [10.164.1.10]:5701 - b6e2d4be-6234-4937-80c6-68167d746a98 this
]
```

Is your predefined ClusterIP in the service address range?

Could you provide more information about your setup and deployment steps?

**Comment by Anumita** [comment]
Hi @hasancelik , it was setup issue. I created a different setup using aks-engine with calico and wasn't facing this error. I am facing another issue though. Each sidecar discovers itself (i.e. gets it's node name) pretty fast, but takes 5 mins to discover the other members. Not sure why this could be happening. I removed the go-app container and  created the replicas with just the hazelcast container and it discovered the members instantly. Not sure why the go-client is creating issues here. Should i raise another issue for this?
